linalg/constraints.hpp

```diff
@@ -0,0 +1,404 @@
+// Copyright (c) 2010-2021, Lawrence Livermore National Security, LLC. Produced
+// at the Lawrence Livermore National Laboratory. All Rights reserved. See files
+// LICENSE and NOTICE for details. LLNL-CODE-806117.
+//
+// This file is part of the MFEM library. For more information and source code
+// availability visit https://mfem.org.
+//
+// MFEM is free software; you can redistribute it and/or modify it under the
+// terms of the BSD-3 license. We welcome feedback and contributions, see file
+// CONTRIBUTING.md for details.
+
+#ifndef MFEM_CONSTRAINED
+#define MFEM_CONSTRAINED
+
+#include "solvers.hpp"
+#include "blockoperator.hpp"
+#include "sparsemat.hpp"
+#include "hypre.hpp"
+
+namespace mfem
+{
+
+class ParFiniteElementSpace;
+
+/** @brief An abstract class to solve the constrained system \f$ Ax = f \f$
+    subject to the constraint \f$ B x = r \f$.
+
+    Although implementations may not use the below formulation, for
+    understanding some of its methods and notation you can think of
+    it as solving the saddle-point system
+
+     (  A   B^T  )  ( x )          (  f  )
+     (  B        )  ( lambda )  =  (  r  )
+
+    Not to be confused with ConstrainedOperator, which is totally
```

Comment:
Somewhere, we should include some documentation about how these two similarly named things are different.